### PR TITLE
CASSANDRA-16385

### PR DESCRIPTION
**Pull request by jonmeredith**
See [CASSANDRA-16385](https://issues.apache.org/jira/browse/CASSANDRA-16385)

### CODE REVIEW

File: src/java/org/apache/cassandra/net/Message.java [hunk 1]
```diff
@@ -757,7 +757,11 @@ private Header extractHeaderPost40(ByteBuffer buf, InetAddressAndPort from, long
         {
             serializeHeaderPost40(message.header, out, version);
             out.writeUnsignedVInt(message.payloadSize(version));
-            message.verb().serializer().serialize(message.payload, out, version);
+            IVersionedAsymmetricSerializer<T,T> payloadSerializer = message.header.verb.serializer();
+            // per-response serializers are only required for in-jvm dtests for message filtering
+            if (null == payloadSerializer)
+                payloadSerializer = instance().callbacks.get(message.header.id, message.header.from).responseVerb.serializer();
```
Comment (yifan-c) [inline src/java/org/apache/cassandra/net/Message.java]:
```suggestion
                payloadSerializer = instance().callbacks.responseSerializer(message.header.id, message.header.from);
                // need to perform null check for the payloadSerializer
```

There is an existing API in `RequestCallbacks` that does the same thing. 

The returned `serializer` could be null. Need to check. If null, it should skip ser/deser. Same for the other methods changed in this PR.

**Review by yifan-c — COMMENTED** [review]
As Alex mentioned, it feels a bit unnecessary to add the per-response serializer code just for in-jvm dtests. 

How about this? 
1. Add a new parameter `IVersionedAsymmetricSerializer<In, Out> serializer` to `serializePost40()`. Same for the other methods. 
2. Add method `serialize(Message<T> message, DataOutputPlus out, int version, IVersionedAsymmetricSerializer<In, Out> serializer)` that is annotated with `@VisibleForTesting`.

The in-jvm dtest framework can determine the serializer to use and call the new testing only `serialize()` method.

**Comment by jonmeredith** [comment]
Efficiency wise I'm not sure all the extra argument passing is better than a null check. I'll see what I can do to improve it. Thanks for the feedback.

**Comment by yifan-c** [comment]
> Efficiency wise I'm not sure all the extra argument passing is better than a null check. I'll see what I can do to improve it. Thanks for the feedback.

They are not about solving the same problem. 

The extra argument and the testing-only method is to separate out the testing code in `Message.java`.

The null check is needed as long as response serializer is fetched. The response serializer can be null if the message has expired.

File: test/distributed/org/apache/cassandra/distributed/upgrade/UpgradeTestBase.java [hunk 2]
```diff
@@ -177,6 +178,10 @@ public void run() throws Throwable
                             cluster.get(n).shutdown().get();
                             cluster.get(n).setVersion(version);
                             cluster.get(n).startup();
+                            // If using in-jvm Dtest networking, update the messaging versions
+                            // so that message filters are able to serialize/deserialize correctly.
+                            if (!cluster.get(n).config().has(Feature.NETWORK))
+                                cluster.updateMessagingVersions();
```
Comment (yifan-c) [inline test/distributed/org/apache/cassandra/distributed/upgrade/UpgradeTestBase.java]: Looks like `updateMessagingVersions()` is already called during `cluster.get(n).startup()`. 

```
AbstractCluster.Wrapper#startup(org.apache.cassandra.distributed.api.ICluster)
```

Do we need to invoke it twice?

File: test/distributed/org/apache/cassandra/distributed/upgrade/UpgradeTest.java [hunk 3]
```diff
@@ -20,13 +20,48 @@
...
 
+import org.apache.cassandra.distributed.UpgradeableCluster;
 import org.apache.cassandra.distributed.api.ConsistencyLevel;
 import org.apache.cassandra.distributed.api.Feature;
 import org.apache.cassandra.distributed.shared.Versions;
 import static org.apache.cassandra.distributed.shared.AssertUtils.*;
 
 public class UpgradeTest extends UpgradeTestBase
 {
+    public static void truncateWriteReadTest(UpgradeableCluster cluster, int instanceNum)
```
Comment (yifan-c) [inline test/distributed/org/apache/cassandra/distributed/upgrade/UpgradeTest.java]: What does the `truncateWriteReadTest` test? Is it for testing the mix-mode messaging filter?

File: src/java/org/apache/cassandra/net/RequestCallbacks.java [hunk 4]
```diff
@@ -116,7 +116,9 @@ public void addWithExpiration(AbstractWriteResponseHandler<?> cb,
         assert previous == null : format("Callback already exists for id %d/%s! (%s)", message.id(), to.endpoint(), previous);
     }
 
-    <T> IVersionedAsymmetricSerializer<?, T> responseSerializer(long id, InetAddressAndPort peer)
+    @Deprecated // for 3.0 compatibility purposes only
```
Comment (yifan-c) [inline src/java/org/apache/cassandra/net/RequestCallbacks.java:119]: 👍 to adding the `@Deprecated`

File: src/java/org/apache/cassandra/net/Message.java [hunk 5]
```diff
@@ -914,9 +922,14 @@ private Header extractHeaderPre40(ByteBuffer buf, long currentTimeNanos, int ver
 
             if (message.payload != null && message.payload != NoPayload.noPayload)
             {
+                IVersionedAsymmetricSerializer<T,T> payloadSerializer = message.header.verb.serializer();
+                // per-response serializers are only required for in-jvm dtests for message filtering
+                if (null == payloadSerializer)
+                    payloadSerializer = instance().callbacks.get(message.header.id, message.header.from).responseVerb.serializer();
+
```
Comment (yifan-c) [inline src/java/org/apache/cassandra/net/Message.java]: Can we also pass the payloadSerializer down to this method?

File: src/java/org/apache/cassandra/net/Message.java [hunk 6]
```diff
@@ -1285,10 +1298,18 @@ private int extractParamsSizePre40(ByteBuffer buf, int readerIndex, int readerLi
...
+                IVersionedAsymmetricSerializer<T,T> payloadSerializer = message.verb().serializer();
+                if (null == payloadSerializer) // per-response serializers are only required for in-jvm dtests for message filtering
+                    payloadSerializer = instance().callbacks.get(message.header.id, message.header.from).responseVerb.serializer();
+                long payloadSize = payloadSerializer.serializedSize(message.payload, version);
+                return Ints.checkedCast(payloadSize);
+            }
+            else
+            {
+                return 0;
+            }
```
Comment (yifan-c) [inline src/java/org/apache/cassandra/net/Message.java]: Can we also pass the payloadSerializer down to this method?

**Comment by yifan-c** [comment]
How does this sound to you? https://github.com/yifan-c/cassandra/commit/9f9498abb96fc0d05417ca6f8c74b180a90ebb1f

File: test/distributed/org/apache/cassandra/distributed/upgrade/UpgradeTestBase.java (hunk 2 above)
Comment (jonmeredith) [reply test/distributed/org/apache/cassandra/distributed/upgrade/UpgradeTestBase.java]: I'd missed that and made this change to make sure the messaging version was set on upgrade. I added the guard around the NETWORK feature so as not to affect any version negotiation attempts if using real networking.

It's probably safest to leave it alone. I'll remove from my code.

File: test/distributed/org/apache/cassandra/distributed/upgrade/UpgradeTest.java (hunk 3 above)
Comment (jonmeredith) [reply test/distributed/org/apache/cassandra/distributed/upgrade/UpgradeTest.java]: Apologies, it was from an earlier iteration where I was trying to prove the issue but it turns out truncate doesn't check the response so it didn't hit the error so I switched to using CAS  for the `Pre40MessageFilterTest`. I'll remove it.

File: src/java/org/apache/cassandra/net/Message.java (hunk 6 above)
Comment (jonmeredith) [reply src/java/org/apache/cassandra/net/Message.java]: Yes, thanks for the patch, that's how it should be.

**Comment by jonmeredith** [comment]
Thanks for the feedback and fixup @yifan-c

File: test/distributed/org/apache/cassandra/distributed/upgrade/UpgradeTest.java [hunk 7]
```diff
@@ -37,21 +38,10 @@ public void upgradeTest() throws Throwable
...
-                assertRows(cluster.coordinator(1).execute("SELECT * FROM " + KEYSPACE + ".tbl WHERE pk = ?",
-                                                          ConsistencyLevel.ALL,
-                                                          1),
-                           row(1, 1, 1),
-                           row(1, 2, 2),
-                           row(1, 3, 3));
-            }
-        }).run();
+        .runAfterNodeUpgrade(UpgradeTest::truncateWriteReadTest)
+        .run();
```
Comment (yifan-c) [inline test/distributed/org/apache/cassandra/distributed/upgrade/UpgradeTest.java:44]: Need to revert the change after removing `truncateWriteReadTest()`

File: test/distributed/org/apache/cassandra/distributed/impl/Instance.java [hunk 8]
```diff
@@ -318,7 +320,11 @@ private static IMessage serializeMessage(InetAddressAndPort from, InetAddressAnd
...
         try (DataOutputBuffer out = new DataOutputBuffer(1024))
         {
-            Message.serializer.serialize(messageOut, out, toVersion);
+            IVersionedAsymmetricSerializer<?,?> payloadSerializer = messageOut.header.verb.serializer();
+            if (null == payloadSerializer)
+                payloadSerializer = MessagingService.instance().callbacks.responseSerializer(messageOut.header.id, messageOut.header.from);
+
+            Message.serializer.serialize(messageOut, out, toVersion, (IVersionedAsymmetricSerializer) payloadSerializer);
             byte[] bytes = out.toByteArray();
             if (messageOut.serializedSize(toVersion) != bytes.length)
```
Comment (yifan-c) [inline test/distributed/org/apache/cassandra/distributed/impl/Instance.java:329]:
```suggestion
            int serializedSize = messageOut.serializedSize(toVersion, (IVersionedAsymmetricSerializer) payloadSerializer);
            if (serializedSize != bytes.length)
                throw new AssertionError(String.format("Message serializedSize(%s) does not match what was written with serialize(out, %s) for verb %s and serializer %s; " +
                                                       "expected %s, actual %s", toVersion, toVersion, messageOut.verb(), messageOut.serializer.getClass(),
                                                       serializedSize, bytes.length));
```
Sorry that I forgot to update the usage of `serializedSize()`. We need to pass in the `payloadSerializer`

**Comment by jonmeredith** [comment]
Merged alternate version as  838bd8cae6d125cfb055dbf6bfc5a2d4e13cdfcb instead.
